tools/crm_resource.c

```diff
@@ -684,6 +684,528 @@ static pcmk__cli_option_t long_options[] = {
+                                            options.rsc_id, options.host_uname,
+                                            options.promoted_role_only);
+
+    } else if (options.host_uname) {
+        dest = pe_find_node(data_set->nodes, options.host_uname);
+        if (dest == NULL) {
+            rc = pcmk_rc_node_unknown;
+            if (BE_QUIET == FALSE) {
+                g_list_free(before);
+            }
+            return rc;
+        }
+        rc = cli_resource_clear(options.rsc_id, dest->details->uname, NULL, cib_conn, TRUE);
+
+    } else {
+        rc = cli_resource_clear(options.rsc_id, NULL, data_set->nodes, cib_conn, TRUE);
+    }
+
+    if (BE_QUIET == FALSE) {
+        rc = cib_conn->cmds->query(cib_conn, NULL, cib_xml_copy, cib_scope_local | cib_sync_call);
+        rc = pcmk_legacy2rc(rc);
+
+        if (rc != pcmk_rc_ok) {
+            g_set_error(&error, PCMK__RC_ERROR, rc,
+                        "Could not get modified CIB: %s\n", pcmk_strerror(rc));
+            g_list_free(before);
+            return rc;
+        }
+
+        data_set->input = *cib_xml_copy;
+        cluster_status(data_set);
+
+        after = build_constraint_list(data_set->input);
+        remaining = subtract_lists(before, after, (GCompareFunc) strcmp);
+
+        for (ele = remaining; ele != NULL; ele = ele->next) {
+            printf("Removing constraint: %s\n", (char *) ele->data);
+        }
+
+        g_list_free(before);
+        g_list_free(after);
+        g_list_free(remaining);
+    }
+
+    return rc;
+}
+
+static int
+delete()
+{
+    int rc = pcmk_rc_ok;
+    xmlNode *msg_data = NULL;
+
+    if (options.rsc_type == NULL) {
+        rc = ENXIO;
+        g_set_error(&error, PCMK__RC_ERROR, rc,
+                    "You need to specify a resource type with -t");
+        return rc;
+    }
+
+    msg_data = create_xml_node(NULL, options.rsc_type);
+    crm_xml_add(msg_data, XML_ATTR_ID, options.rsc_id);
+
+    rc = cib_conn->cmds->remove(cib_conn, XML_CIB_TAG_RESOURCES, msg_data, cib_options);
+    rc = pcmk_legacy2rc(rc);
+    free_xml(msg_data);
+    return rc;
+}
+
+static int
+list_agents(const char *spec, crm_exit_t *exit_code)
+{
+    int rc = pcmk_rc_ok;
+    lrmd_list_t *list = NULL;
+    lrmd_list_t *iter = NULL;
+    char *provider = strchr (spec, ':');
+    lrmd_t *lrmd_conn = lrmd_api_new();
+
+    if (provider) {
+        *provider++ = 0;
+    }
+    rc = lrmd_conn->cmds->list_agents(lrmd_conn, &list, spec, provider);
+
+    if (rc > 0) {
+        for (iter = list; iter != NULL; iter = iter->next) {
+            printf("%s\n", iter->val);
+        }
+        lrmd_list_freeall(list);
+        rc = pcmk_rc_ok;
+    } else {
+        *exit_code = CRM_EX_NOSUCH;
+        g_set_error(&error, PCMK__EXITC_ERROR, *exit_code,
+                    "No agents found for standard=%s, provider=%s",
+                    spec, (provider? provider : "*"));
+    }
+
+    lrmd_api_delete(lrmd_conn);
+    return rc;
```

Comment:
In the error case this is still a legacy rc (similar below)

Reply:
This is fixed in another patch later in the series, but I've corrected it so everything should be in the right place now.